plan/column_pruning.go

```diff
@@ -358,3 +358,41 @@ func (p *Apply) PruneColumnsAndResolveIndices(parentUsedCols []*expression.Colum
 	p.schema.InitIndices()
 	return append(childOuterUsedCols, p.outerColumns...), nil
 }
+
+// PruneColumnsAndResolveIndices implements LogicalPlan PruneColumnsAndResolveIndices interface.
+// NewUpdate do not prune columns. Here we just do two things:
+// 1. resolve indices for schema
+// 2. reorder OrderedList
+func (p *NewUpdate) PruneColumnsAndResolveIndices(parentUsedCols []*expression.Column) ([]*expression.Column, error) {
+	outer, err := p.baseLogicalPlan.PruneColumnsAndResolveIndices(p.GetSchema())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// column prune may reorder schema, so we re-evaluate p.Orderedlist
+	orderedList := make([]*expression.Assignment, len(p.OrderedList))
+	for _, v := range p.OrderedList {
+		if v == nil {
+			continue
+		}
+		orderedList[p.GetSchema().GetIndex(v.Col)] = v
+	}
+
+	for i := 0; i < len(orderedList); i++ {
+		if orderedList[i] == nil {
+			continue
+		}
+		orderedList[i].Col.Index = p.GetSchema().GetIndex(orderedList[i].Col)
+		switch assign := orderedList[i].Expr; assign.(type) {
+		case (*expression.Column):
+			assign.(*expression.Column).Index = p.GetSchema().GetIndex(assign.(*expression.Column))
+		case (*expression.ScalarFunction):
+			for i, args := 0, assign.(*expression.ScalarFunction).Args; i < len(args); i++ {
```

Comment:
Column may not be direct argument of the scalar function, like `a = 3 * (a + 1)`